### ISSUE TITLE
ArrayObject Question

### ISSUE THREAD

**Issue by designermonkey** [ryancramerdesign/ProcessWire#1501]
Hi there. I have a question about adding ArrayObject style access to ProcessWire.

I've had a quick look at how PW does it's thing, and can see that a lot of properties are publicly available, but everything is dependant on using object style access to a Page instance for example.

The kind of work I do means that an API that allowed me to have ArrayAccess alongside object style is a requirement, and also I believe, a benefit. I've made a class in other projects that allows this so I can use my data objects (like PWs Page) directly inside a template engine like Mustache (where mustache-php requires array access).

Being new to PW, I am just wondering how big a job you think it would be to add this level of functionality in? To be very useable, an iterable version would be needed (for PageArray for example) and a non iterable version for pages themselves.

I'm still at the beginning of looking into how the class hierarchy works, but I think the level would be down at the Wire class and WireArray class using traits to add specific functionality.

So, the main question... Do you think this is something that PW would adopt? Would you be willing to do it?

**Comment by ryancramerdesign** [comment]
Page objects themselves don't support ArrayAccess, but I'd be happy to add it if it would help for something like mustache. Admittedly, I tend to avoid use of ArrayAccess just because I like to have a clear distinction between what are PHP arrays and what are PHP objects. But that's only personal preference, and I don't see any harm in adding it, because one isn't required to use it. 

Page objects are already iterable with the IteratorAggregate interface. If you iterate a Page, you'll get all of it's populated settings and fields. Since PW loads fields on-demand, as they are needed, you do impart  unnecessary load and memory usage if you start using this for all pages though. So I'd only recommend iterating a Page when you already know you are going to need all of its data.

PageArray (and all WireArray derived objects) already implement ArrayAccess and IteratorAggregate. You don't see us using ArrayAccess for WireArrays/PageArrays in any of our examples or documentation, but it is there and supported.

**Comment by designermonkey** [comment]
Hi, thanks for the feedback.

I think the only issue would be the iterability of some of the objects then, as I know mustache-php checks for this to distinguish between associative arraylikes (non-iterable) and keyed arraylikes (iterable).

**Comment by ryancramerdesign** [comment]
I'm not sure I understand that aspect of Muscache, as associative arrays are iterable in PHP so not sure why they shouldn't be as objects. There are many parts in PW that depend on the ability to iterate objects that use associative keys. Since it's a common usage (whether in PW or elsewhere) I'm thinking there must be some setting in Mustache that lets you configure that behavior? At minimum, it seems like they should at least check if the key is a string before deciding whether it's associative or non. If not, it sounds like a limitation of Mustache. On our side, I'm good to add support for things that weren't there (like ArrayAccess) but wouldn't want to start removing important and useful things to support a limitation in an external library. 

Btw I have updated the base WireData class in PW 3.x (devns branch) to support ArrayAccess, and since the Page objects are a type of WireData object, they are included too. I'm guessing there's a simple solution for the other Mustache issue, but I'm not familiar enough with that project to know what it might be.

**Comment by designermonkey** [comment]
Yeah, it was a confusion for me too, but we have to remember that Mustache isn't just PHP, and applies to other languages, which use 'hashes'. In a hash, a keyed object is an object, not an array, and should not be traversable. It's the only way to distinguish between an array and a 'context'; a context being dot separated.

It's a feature parity thing, nothing more, but it can be a little confusing (it was for me).

Thanks for adding the ArrayAccess. When I get some time I'll have a proper play with PW and see what I can come up with.
